R/future_promise.R

```diff
@@ -0,0 +1,500 @@
+    },
+
+    # Returns a logical which let's work begin
+    can_proceed = function() {
+      isTRUE(private$can_proceed_fn())
+    },
+
+    # Function to attempt as much work as possible
+    # If no workers are available and a queue has elements,
+    #   If a delayed check has already been registered, Return
+    #   Else, check again after some delay
+    attempt_work = function(can_delay = FALSE) {
+      debug_msg('attempt_work()')
+      # If nothing to start, return early
+      if (private$queue$size() == 0) return()
+
+      # If we are not waiting on someone else, we can do work now
+      while ((private$queue$size() > 0) && private$can_proceed()) {
+        # Do work right away
+        private$reset_delay()
+        private$do_work()
+      }
+
+      # If there are still items to be processed, but we can not proceed...
+      if (private$queue$size() > 0 && ! private$can_proceed()) {
+        # If we are allowed to delay (default FALSE), or nothing is currently delaying
+        if (can_delay || is.null(private$cancel_delayed_attempt_work)) {
+
+          # Try again later
+          private$increase_delay()
+          private$cancel_delayed_attempt_work <-
+            later::later(
+              loop = private$loop,
+              delay = private$delay$delay(),
+              function() {
+                private$attempt_work(can_delay = TRUE)
+              }
+            )
+        }
+      }
+    },
+
+    # Actually process an item in the queue
+    do_work = function() {
+      debug_msg("do_work()")
+
+      # Get first item in queue
+      work_fn <- private$queue$remove()
+
+      # Safety check...
+      # If nothing is returned, no work to be done. Return early
+      if (!is.function(work_fn)) return()
+
+      # Do scheduled work
+      debug_msg("execute work")
+      future_job <- work_fn()
+
+      # Try to attempt work immediately after the future job has finished
+      finally(future_job, function() {
+        debug_msg("finished work. queue size: ", private$queue$size())
+        private$attempt_work()
+      })
+
+      return()
+    }
+  ),
+  public = list(
+    #' @description Create a new `WorkQueue`
+    #' @param can_proceed Function that should return a logical value. If `TRUE` is returned, then the next scheduled work will be executed. By default, this function checks if \code{\link[future:nbrOfWorkers]{future::nbrOfFreeWorkers()} > 0}
+    #' @param queue Queue object to use to store the scheduled work. By default, this is a "First In, First Out" queue using [fastmap::fastqueue()]. If using your own queue, it should have the methods `$add(x)`, `$remove()`, `$size()`.
+    #' @param loop \pkg{later} loop to use for calculating the next delayed check. Defaults to [later::global_loop()].
+    initialize = function(
+      # defaults to a future::plan agnostic function
+      can_proceed = future_worker_is_free,
+      queue = fastmap::fastqueue(), # FIFO
+      loop = later::global_loop()
```

Comment:
Add a comment about why we're currently using global, why it might not work for private event loops that are meant to be synchronized, and that we might need to change the behavior in the future, but aren't sure exactly how at this point. Having an actual use case will help us understand the problem better.